spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java

```diff
@@ -0,0 +1,108 @@
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.iceberg.Files;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.orc.ORC;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+import org.apache.orc.OrcFile;
+import org.apache.orc.TypeDescription;
+import org.apache.orc.Writer;
+import org.apache.orc.storage.ql.exec.vector.BytesColumnVector;
+import org.apache.orc.storage.ql.exec.vector.LongColumnVector;
+import org.apache.orc.storage.ql.exec.vector.UnionColumnVector;
+import org.apache.orc.storage.ql.exec.vector.VectorizedRowBatch;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+
+public class TestSparkOrcUnions {
+  private static final int BATCH_SIZE = 50;
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  @Test
+  public void testComplexUnion() throws IOException {
+    TypeDescription orcSchema =
+        TypeDescription.fromString("struct<unionCol:uniontype<int,string>>");
+
+    Schema expectedSchema = new Schema(
+        Types.NestedField.optional(0, "unionCol", Types.StructType.of(
+            Types.NestedField.optional(1, "tag_0", Types.IntegerType.get()),
+            Types.NestedField.optional(2, "tag_1", Types.StringType.get())))
+    );
+
+    Configuration conf = new Configuration();
+
+    temp.delete();
+    String basePath = temp.getRoot().getAbsolutePath();
+    Path path = new Path(basePath + "/test.orc");
+
+    Writer writer = OrcFile.createWriter(path,
+        OrcFile.writerOptions(conf)
+            .setSchema(orcSchema));
+
+    VectorizedRowBatch batch = orcSchema.createRowBatch();
+    LongColumnVector longColumnVector = new LongColumnVector(BATCH_SIZE);
+    BytesColumnVector bytesColumnVector = new BytesColumnVector(BATCH_SIZE);
+    UnionColumnVector complexUnion = new UnionColumnVector(BATCH_SIZE, longColumnVector, bytesColumnVector);
+    complexUnion.init();
+
+    for (int i = 0; i < BATCH_SIZE; i += 1) {
+      complexUnion.tags[i] = i % 2;
+      longColumnVector.vector[i] = i;
+      String stringValue = "stringtype" + i;
+      bytesColumnVector.setVal(i, stringValue.getBytes(StandardCharsets.UTF_8));
+    }
+
+    batch.size = BATCH_SIZE;
+    batch.cols[0] = complexUnion;
+
+    writer.addRowBatch(batch);
+    batch.reset();
+    writer.close();
+
+    List<InternalRow> internalRows = Lists.newArrayList();
+    try (CloseableIterable<InternalRow> reader = ORC
+        .read(Files.localInput(new File(path.getParent() + "/" + path.getName())))
+        .project(expectedSchema)
+        .createReaderFunc(readOrcSchema -> new SparkOrcReader(expectedSchema, readOrcSchema))
+        .build()) {
+      reader.forEach(internalRows::add);
+
+      Assert.assertEquals(internalRows.size(), BATCH_SIZE);
```

Comment:
Can we also assert the content of the rows read in addition to just checking the number of rows read?